codec-http/src/test/java/io/netty/handler/codec/http/HttpRequestEncoderTest.java

```diff
@@ -182,6 +181,180 @@ public void testEmptyContentNotsChunkedWithTrailers() throws Exception {
+            return content;
+        }
+
+        @Override
+        public int refCnt() {
+            return content.refCnt();
+        }
+
+        @Override
+        public CustomFullHttpRequest retain() {
+            content.retain();
+            return this;
+        }
+
+        @Override
+        public CustomFullHttpRequest retain(int increment) {
+            content.retain(increment);
+            return this;
+        }
+
+        @Override
+        public CustomFullHttpRequest touch() {
+            content.touch();
+            return this;
+        }
+
+        @Override
+        public CustomFullHttpRequest touch(Object hint) {
+            content.touch(hint);
+            return this;
+        }
+
+        @Override
+        public boolean release() {
+            return content.release();
+        }
+
+        @Override
+        public boolean release(int decrement) {
+            return content.release(decrement);
+        }
+
+        @Override
+        public CustomFullHttpRequest setProtocolVersion(HttpVersion version) {
+            super.setProtocolVersion(version);
+            return this;
+        }
+
+        @Override
+        public CustomFullHttpRequest setMethod(HttpMethod method) {
+            super.setMethod(method);
+            return this;
+        }
+
+        @Override
+        public CustomFullHttpRequest setUri(String uri) {
+            super.setUri(uri);
+            return this;
+        }
+
+        @Override
+        public FullHttpRequest copy() {
+            return replace(content().copy());
+        }
+
+        @Override
+        public FullHttpRequest duplicate() {
+            return replace(content().duplicate());
+        }
+
+        @Override
+        public FullHttpRequest retainedDuplicate() {
+            return replace(content().retainedDuplicate());
+        }
+
+        @Override
+        public FullHttpRequest replace(ByteBuf content) {
+            FullHttpRequest request = new DefaultFullHttpRequest(protocolVersion(), method(), uri(), content,
+                    headers().copy(), trailingHeaders().copy());
+            request.setDecoderResult(decoderResult());
+            return request;
+        }
+
+        @Override
+        public int hashCode() {
+            int hash = this.hash;
+            if (hash == 0) {
+                if (ByteBufUtil.isAccessible(content())) {
+                    try {
+                        hash = 31 + content().hashCode();
+                    } catch (IllegalReferenceCountException ignored) {
+                        // Handle race condition between checking refCnt() == 0 and using the object.
+                        hash = 31;
+                    }
+                } else {
+                    hash = 31;
+                }
+                hash = 31 * hash + trailingHeaders().hashCode();
+                hash = 31 * hash + super.hashCode();
+                this.hash = hash;
+            }
+            return hash;
```

Comment:
How come we need to go to these lengths for the equals and hashCode?

Reply:
Good point, no we don't need it indeed :O 
let me simplify it

Reply:
Ah no, I've copied from default full http msg I see :O

Reply:
So what's the conclusion?

Reply:
It was copied by `DefaultFullHttpMessage` like the most of its code!